### DISCUSSION TITLE
Scrollback state determin.

### DISCUSSION THREAD

**Discussion by firedes** [wezterm/wezterm#3641] [Q&A] [answered]
Hi, I'd like to do some vim like bindings in wezterm, so the `CTRL+F` is sure to be scroll page down, but `CTRL+F` is a useful keybinding in my zsh environment which is forward word for completion at input line? After some research I figured out that I can map `CTRL+F` to use `sendKey` when I'm at command input line, and perform `ScrollByPage=1` when I have done any scroll page up action. is this possible? I did a cursory look at the docs, but couldn't find any API that can determine the scrollback state.

**Comment by yutkat** [comment]
If you write a config like this, you can call the wezterm command from the CLI.

```lua
local hacky_user_commands = {
	["scroll-up"] = function(window, pane, cmd_context)
		window:perform_action(wezterm.action({ ScrollByPage = -1 }), pane)
		-- wezterm.action({ ScrollByPage = -1 })
	end,
	["scroll-down"] = function(window, pane, cmd_context)
		window:perform_action(wezterm.action({ ScrollByPage = 1 }), pane)
	end,
}

wezterm.on("user-var-changed", function(window, pane, name, value)
	if name == "hacky-user-command" then
		local cmd_context = wezterm.json_parse(value)
		hacky_user_commands[cmd_context.cmd](window, pane, cmd_context)
		return
	end
end)
```

If you could run this command if there is no input at the zsh prompt, you could do what you are hoping to do. (I don't know if it would be useful).

```bash
printf "\033]1337;SetUserVar=%s=%s\007" hacky-user-command $(printf '{"cmd":"scroll-up"}' | base64)
```

By the way, I use alt-f/b for scroll-up/down. Also, if you want to operate shell that much like a vim,  you might want to use zsh's vim mode.

https://github.com/wez/wezterm/issues/2979#issuecomment-1447519267

**Reply by firedes** [reply]
Sorry for my bad English. Definitely I'm using zsh's vim mode. I'm not asking something like making wezterm's functionalities be using in shell cli. What I'm actually asking is something like key binding conflict:

```lua
{
  key = "f",
  mods = "CTRL",
  action = wezterm.action_callback(function(win, pane)
    if is_vim(pane) then
      win:perform_action({ SendKey = { key = "f", mods = "CTRL" } }, pane)
    elseif (???) then
      -- We have not scrolled page up yet, so ctrl-f should be SendKey to the cli directly,
      -- so that it can function as some other key binding in zsh vi mode(in my case, forward word for command completion).
      win:perform_action({ SendKey = { key = "f", mods = "CTRL" } }, pane)
    else
      win:perform_action(act({ ScrollByPage = 1 }), pane)
    end
  end),
},

```
```bash
~  bindkey-all|grep -F "^F"
#### .safe
#### command
#### emacs
#### isearch
#### main
#### vicmd
#### viins
#### viopp
#### visual
"^F" forward-char
"^X^F" vi-find-next-char
"^F" vi-forward-word
"^F" vi-forward-word
```

**Reply by yutkat** [reply]
I think you are trying to control zsh from wezterm, but it would be better to do the opposite, to control wezterm from zsh. Because it is difficult to know the status of zsh.

**Reply by firedes** [reply]
I don't think it's difficult for lua layer to know the status of scrolling, think about if an api about scroll bar including its position info exposed by wezterm. Anyway, I'm currently using a tricky one to solve this. Appreciate for your patience and suggestion.
```lua
{
  key = "f",
  mods = "CTRL",
  action = wezterm.action_callback(function(win, pane)
    if is_vim(pane) then
      win:perform_action({ SendKey = { key = "f", mods = "CTRL" } }, pane)
    else
      win:perform_action(act({ ScrollByPage = 1 }), pane)
      win:perform_action({ SendKey = { key = "f", mods = "CTRL" } }, pane)
    end
  end),
}
```

**Reply by yutkat** [reply]
If there is the scrollbar position API, you could see whether or not it is in a scrolling state from the position of the scrollbar. But I don't think `ctrl-b` would work with that strategy.

If all you need is `ctrl-f`, that hack may indeed work.

Personally, I'd still recommend separating the key bindings instead of having multiple meanings.

**Reply by yutkat** [reply, marked as answer]
If you do it this way, you can make it work for ctrl-b.

```zsh
function ctrl_b_with_wezterm() {
	if [[ -n $BUFFER ]]; then
		zle backward-char
	else
		printf "\033]1337;SetUserVar=%s=%s\007" hacky-user-command $(printf '{"cmd":"scroll-up"}' | base64)
	fi
}
zle -N ctrl_b_with_wezterm
bindkey '^B' ctrl_b_with_wezterm
```

However, you need to disable this option (I don't use this method because this option is important to me)

```lua
scroll_to_bottom_on_input = false,
```

**Reply by firedes** [reply]
Thanks, a very useful idea.
